packages/core-ethereum/crates/core-ethereum-rpc/src/rpc.rs

```diff
@@ -0,0 +1,326 @@
+use async_trait::async_trait;
+use core_crypto::keypairs::{ChainKeypair, Keypair};
+use core_crypto::types::Hash;
+use core_ethereum_types::{ContractAddresses, ContractInstances};
+use ethers::middleware::{MiddlewareBuilder, NonceManagerMiddleware, SignerMiddleware};
+use ethers::prelude::k256::ecdsa::SigningKey;
+use ethers::prelude::transaction::eip2718::TypedTransaction;
+use ethers::signers::{LocalWallet, Signer, Wallet};
+use ethers::types::BlockId;
+use ethers_providers::{JsonRpcClient, Middleware, Provider, RetryClient, RetryClientBuilder, RetryPolicy};
+use primitive_types::H160;
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+use std::time::Duration;
+use utils_types::primitives::{Address, Balance, BalanceType, U256};
+use validator::Validate;
+
+use crate::errors::Result;
+use crate::HoprRpcOperations;
+
+#[derive(Clone, Debug, Copy, PartialEq, Eq, Serialize, Deserialize, Validate)]
+pub struct RpcOperationsConfig {
+    pub chain_id: u64,
+    pub contract_addrs: ContractAddresses,
+    pub max_http_retries: u32,
+    pub expected_block_time: Duration,
+    pub logs_page_size: u64,
+    pub tx_polling_interval: Duration,
+}
+
+impl Default for RpcOperationsConfig {
+    fn default() -> Self {
+        Self {
+            chain_id: 100,
+            contract_addrs: Default::default(),
+            max_http_retries: 3,
+            logs_page_size: 50,
+            expected_block_time: Duration::from_secs(5),
+            tx_polling_interval: Duration::from_secs(7),
+        }
+    }
+}
+
+pub(crate) type HoprMiddleware<P> =
+    NonceManagerMiddleware<SignerMiddleware<Provider<RetryClient<P>>, Wallet<SigningKey>>>;
+
+#[derive(Debug, Clone)]
+pub struct RpcOperations<P: JsonRpcClient + 'static> {
+    me: Address,
+    pub(crate) provider: Arc<HoprMiddleware<P>>,
+    pub(crate) cfg: RpcOperationsConfig,
+    contract_instances: ContractInstances<HoprMiddleware<P>>,
+}
+
+impl<P: JsonRpcClient + 'static> RpcOperations<P> {
+    pub fn new<R>(json_rpc: P, chain_key: &ChainKeypair, cfg: RpcOperationsConfig, retry_policy: R) -> Result<Self>
+    where
+        R: RetryPolicy<<P as JsonRpcClient>::Error> + 'static,
+    {
+        let provider_client = RetryClientBuilder::default()
+            .rate_limit_retries(5)
```

Comment:
We snould make these values configurable.

Reply:
This one should be the same as the one after it, because when using other client than `Reqwest` (which is the case when non-native), the RetryClient applies the `rate_limit_retries` also to timeout errors.